### PR TITLE
route for object introspection

### PR DESCRIPTION
**Pull request by sapieneptus**
Created a route for object introspection. 

Returns JSON in the form of : 
```
{
    methods :  [  <method>, ... ],
    properties: {
        <propname> : { setter : <setter name>, getter: <getter name> }
    }
}

### CODE REVIEW

**Comment by jmoody** [comment]
Jenkins please test this.

File: calabash/Classes/Utils/LPJSONUtils+Introspection.m [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+ do not follow the custom naming conventions. E.g., for property
+ 'foo', the custom naming conventions would be getter=foo, setter=setFoo.
+ If the names are different than convention dictates (e.g. getter=obtainFooValue,
+ setter=updateFooValue), they are included in the returned list.
+ 
+ @param object - the object for which you'd like the custom accessors
+ @return NSDictionary - a dictionary of strings describing the custom accessors, or nil
+ if none exist.
+ */
+
```
Comment (jmoody) [inline calabash/Classes/Utils/LPJSONUtils+Introspection.m]: Prefer no newline here.

File: calabash/Classes/Utils/LPJSONUtils+Introspection.m [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        break;
+        
+      case 'R': /* readonly */
+        readonly = YES;
+        accessors[@"setter"] = @"READONLY";
+        break;
+        
+      default:
+        break;
+    }
```
Comment (jmoody) [inline calabash/Classes/Utils/LPJSONUtils+Introspection.m]: Nice.

**Comment by jmoody** [comment]
@sapieneptus 

Looks good.

I am willing to merge as-is.

I have a couple of questions and comments.

* Can the Introspection Category be managed by ARC?
* Is Introspection a good name?
* Why a Category and not a new class?

I really dislike the top matter that Xcode injects into new source files.

```
+//
+//  LPJSONUtils+Accessors.h
+//  calabash
+//
+//  Created by Chris Fuentes on 6/5/15.
+//  Copyright (c) 2015 Xamarin. All rights reserved.
+//
```

Comments like this are from a different century, are essentially meaningless, and resist refactoring.  We cannot remove existing comments like this, but I would rather that no new comments like this be added.

**Comment by jmoody** [comment]
@rasmuskl @john7doe @krukow 

I need some feedback before I merge.  Is this feature desirable?

File: XCTest/Tests/Routes/LPIntrospectionRouteTest.m [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+- (void)testPerformanceExample {
+    // This is an example of a performance test case.
+    [self measureBlock:^{
+        // Put the code you want to measure the time of here.
+    }];
+}
+
+@end
```
Comment (jmoody) [inline XCTest/Tests/Routes/LPIntrospectionRouteTest.m]: This whole block, from line 15 - 40 can be deleted.

File: XCTest/Tests/Routes/LPIntrospectionRouteTest.m [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      }
+      
+      expect(properties[@"string"]).to.equal(@{@"getter" : @"string", @"setter" : @"setString:"});
+      expect(properties[@"readonlyString"]).to.equal(@{@"getter" : @"readonlyString", @"setter" : @"READONLY"});
+      expect(properties[@"customGetterString"]).to.equal(@{@"getter" : @"customStringGetter", @"setter" : @"setCustomGetterString:"});
+      expect(properties[@"customSetterString"]).to.equal(@{@"getter" : @"customSetterString", @"setter" : @"customStringSetter:"});
+      expect(properties[@"customGetterCustomGetterString"]).to.equal(@{@"getter" : @"customStringGetter", @"setter" : @"customStringSetter:"});
+    });
+    
+    it(@"alwaysReturnsAValidSetterAndGetterOrREADONLY", ^{
```
Comment (jmoody) [inline XCTest/Tests/Routes/LPIntrospectionRouteTest.m]: If you prefer, you could write these examples as: @"always return valid setter, getter, or readonly"

File: calabash/Classes/FranklyServer/Routes/LPIntrospectionRoute.m [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+#import "LPIntrospectionRoute.h"
+#import "UIScriptParser.h"
+#import "LPTouchUtils.h"
+#import "LPJSONUtils+Introspection.h"
+
+@interface LPIntrospectionRoute ()
+@end
+
+@implementation LPIntrospectionRoute
```
Comment (jmoody) [inline calabash/Classes/FranklyServer/Routes/LPIntrospectionRoute.m]: Is this missing the:

```
- (BOOL) supportsMethod:(NSString *) method atPath:(NSString *) path;
```

method?

Prompted by your questions about how routing is done, I dug into this a bit.  It seems to me that if this method is not being called by the router, it should be.

File: calabash/Classes/Utils/LPJSONUtils+Introspection.m [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    NSString *theRest         = [propName substringFromIndex:1];
+    NSString *setter          = [NSString stringWithFormat:@"set%@%@:", firstLetterCap, theRest];
+    if ([obj respondsToSelector:NSSelectorFromString(setter)]) { accessors[@"setter"] = setter; }
+    else { accessors[@"setter"] = @"<UNKNOWN>"; }
+  }
+  free(attributes);
+  return accessors;
+}
+
+#define BUF_LEN 256
```
Comment (jmoody) [inline calabash/Classes/Utils/LPJSONUtils+Introspection.m]: I am trying to move away from `#define`, especially those that do not appear at the top of a file.

Does it make sense to make this a static const?

```
static NSUInteger const LPJSONUtilsIntrospectionMethodDescriptionBufferLength = 256;
```

**Comment by jmoody** [comment]
@sapieneptus 

Very nice pull request.  I am thrilled to have another Objective-C nerd looking at this repo.

**Comment by rasmuskl** [comment]
@jmoody Yup. It would enable some nice scenarios where you're able to interactively explore the app in more detail from the REPL instead of diving into docs. It would also allow for better client side error messages for backdoor / invoke scenarios.

File: XCTest/Tests/Routes/LPIntrospectionRouteTest.m (hunk 4 above)
Comment (sapieneptus) [inline XCTest/Tests/Routes/LPIntrospectionRouteTest.m]: Ah, yes that would make sense

**Comment by sapieneptus** [comment]
@jmoody I'll make those few cleanups first. 

Regarding the name, I chose 'introspection' because that's [the vernacular Apple uses] (https://developer.apple.com/library/prerelease/ios/documentation/General/Conceptual/CocoaEncyclopedia/Introspection/Introspection.html) to refer to digging into the details of an object, which is of course the goal of this PR :)

**Comment by sapieneptus** [comment]
Jenkins please test this.

**Comment by sapieneptus** [comment]
Jenkins please test this.

**Comment by jmoody** [comment]
Jenkins test this please

**Comment by sapieneptus** [comment]
Jenkins test this please

**Comment by jmoody** [comment]
@sapieneptus Is this ready to go?

**Comment by sapieneptus** [comment]
@jmoody I believe so

**Pull request merged**